railgun/hack/generators/controllers/networking/main.go

```diff
@@ -234,14 +267,66 @@ var controllerTemplate = `
 type {{.PackageAlias}}{{.Type}}Reconciler struct {
 	client.Client
 
+{{- if or .AcceptsIngressClassNameSpec .AcceptsIngressClassNameAnnotation}}
+	Log              logr.Logger
+	Scheme           *runtime.Scheme
+	Proxy            proxy.Proxy
+	IngressClassName string
+{{- else}}
 	Log    logr.Logger
 	Scheme *runtime.Scheme
 	Proxy  proxy.Proxy
+{{- end}}
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *{{.PackageAlias}}{{.Type}}Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+{{- if .AcceptsIngressClassNameAnnotation}}
+	preds := predicate.NewPredicateFuncs(func(obj client.Object) bool {
```

Comment:
Implementing such (somewhat complex) functions in code generation boilerplate:
- makes the boilerplate harder to read
- makes the boilerplate harder to edit - instead of editing Go code, you edit an opaque huge string in the generator
- makes the logic impossible to unit test

Suggestion: implement and unit test the below function in a non-generated Go package, and then use in a 1-liner fashion in the generated code

`func IngressClassMatcherPredicateFuncs(desiredIngressClass string) predicate.Funcs`

In that case you probably don't need both `AcceptsIngressClassNameAnnotation` and `AcceptsIngressClassNameSpec`. The function above can try match the `ingressClassName` field only on types it expects to have that field (and have a safe failure mode of not matching if the type is not identified as such)

Reply:
To these points:

> makes the boilerplate harder to read

It's true that if you have no way for your editor to selectively identify and syntax highlight these strings differently that is is harder to read.

> makes the boilerplate harder to edit - instead of editing Go code, you edit an opaque huge string in the generator

Also true to some extent.

> makes the logic impossible to unit test

This part is incorrect, it is entirely possible to unit test template logic. I would be in favor of adding the unit tests rather than tearing out the logic, except I do find the readability and edit-ability points above merited so I'm leaning your way.

---

Given your preference and your points about readability/edit-ability I'm inclined to change this as you suggest, I will have a commit up soon :+1:

Reply:
Resolved by 20458030831a8fd09d7d59aabe2a2e3aef4e0151

Reply:
The point has been addressed so the considerations below are purely for academic purposes :mortar_board: 

> This part is incorrect, it is entirely possible to unit test template logic

The template is an opaque string of hundreds of lines and is not code. You can test that the generator generates something that passes unit tests, or you can directly unit test the generated stuff, but in that case you kinda defeat the purpose of generation (by relying on the exact generated content).

Reply:
> The template is an opaque string of hundreds of lines and is not code

I don't follow this assertion, or understand the intention behind it.

Given the Wordnik dictionary definition of "Source Code" (code is very broad):

```
source code
    n.
    Human-readable instructions in a programming language, to be transformed into machine instructions by a compiler, assembler or other translator, or to be carried out directly by an interpreter.
```

Furthermore the [Golang Template Documentation](https://golang.org/pkg/text/template) explicitly [refers](https://golang.org/pkg/text/template/#hdr-Text_and_spaces) to templates as "source" and "code", examples include:

```
"when the template is executed"
```
```
"However, to aid in formatting template source code"
```
```
"only code specified by the template author should run as a result of injecting the template output into a page"
```
What definition of "code" do you prefer that doesn't match Go templates or what am I missing here to better understand the point?

> You can test that the generator generates something that passes unit tests, or you can directly unit test the generated stuff, but in that case you kinda defeat the purpose of generation (by relying on the exact generated content).

I'm not sure I understand what we're getting at here either, if I use the [Agile Definition Of Unit Tests](https://www.agilealliance.org/glossary/unit-test/):

```
[A unit test] is a short program fragment written and maintained by the developers on the product team, which exercises some narrow part of the product’s source code and checks the results. The outcome of a unit test is binary: either “pass” if the program’s behavior is consistent with the recorded expectations, or “fail” otherwise.
```

The template parser when fed the template code can be fed arguments, and given that input produce multiple different outcomes which could either pass or fail `go test` tests requiring only the local system. Why make a distinction?